jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java

```diff
@@ -977,23 +977,27 @@ protected void runJob(Runnable job)
     protected boolean evict()
     {
         long idleTimeoutNanos = TimeUnit.MILLISECONDS.toNanos(getIdleTimeout());
+
+        // There is a chance that many threads enter this method concurrently,
+        // and if all of them are evicted the pool shrinks below minThreads.
+        // For example when minThreads=3, threads=8, maxEvictCount=10 we want
+        // to evict at most 5 threads (8-3), not 10.
+        // When a thread fails the CAS, it may assume that another thread has
+        // been evicted, so the CAS should be attempted only a number of times
+        // equal to the most threads we want to evict (5 in the example above).
+        int threads = getThreads();
+        int minThreads = getMinThreads();
+        int threadsToEvict = threads - minThreads;
+
         while (true)
```

Comment:
why not just:
```
while(threadsToEvict-- > 0)
```
??

Reply:
Because we want to log when there are no more excess threads, and if we put the condition on the `while` becomes more difficult.